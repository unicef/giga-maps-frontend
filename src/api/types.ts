import { GeoJSONGeometry, GeoJSONPoint } from "~/core/global-types";

export type ConnectionStatus = 'no' | 'moderate' | 'good' | 'unknown';

export type GlobalStats = {
  no_of_countries: number;
  schools_connected: number;
  connectivity_global_benchmark: {
    value: number;
    unit: string;
  };
  connected_schools: {
    connected: number;
    not_connected: number;
    unknown: number;
  };
}

export type GraphData = {
  group: string;
  key: string;
  value: number;
}
export type defaultLegendValuesType = {
  good: number;
  moderate: number;
  no_internet: number;
  unknown: number;
}
export type ConnectivityStat = {
  live_avg: number;
  no_of_schools_measure: number;
  school_with_realtime_data: number;
  is_data_synced: boolean;
  real_time_connected_schools: defaultLegendValuesType;
  graph_data: GraphData[];
  live_avg_connectivity: string;
  benchmark_metadata: {
    base_benchmark: string;
    benchmark_unit: string;
    benchmark_value: string;
    parameter_column_unit: string;
    round_unit_value: string;
  }
};
export type SchoolInfoStats = {
  num_students: number;
  num_teachers: number;
  num_classroom: number;
  num_latrines: number;
  running_water: boolean;
  electricity_availability: boolean;
  computer_lab: boolean;
  num_computers: number;
  connectivity: boolean;
  connectivity_status: string;
  connectivity_type: string;
  connectivity_speed: number;
  connectivity_latency: number;
  coverage_availability: boolean;
  coverage_type: string;
  created: string;
  modified: string;
}
export type SchoolStatsType = {
  is_rt_connected: boolean;
  week_connectivity: string;
  id: number;
  name: string;
  field_status?: string;
  field_value?: string;
  coverage_status?: string;
  country_name: string;
  coverage_type: string;
  external_id: string;
  environment: string;
  admin1_name?: string;
  admin2_name?: string;
  admin1_code?: string;
  giga_id_school?: string;
  live_avg_connectivity: string;
  education_level?: string;
  connectivity_speed: number;
  live_avg: number;
  connectivity_status: string;
  connectivity_latency: number;
  admin1_description_ui_label?: string;
  admin2_description_ui_label?: string;
  connectivity_uptime: number;
  statistics: SchoolInfoStats;
  graph_data: GraphData[];
  geopoint: GeoJSONPoint
};


export type IntegrationStatus = 0 | 1 | 2 | 3 | 4 | 5;
export type ConnectivityAvailability =
  | 'no_connectivity'
  | 'connectivity'
  | 'static_speed'
  | 'realtime_speed';

export type CoverageAvailability =
  | 'no_coverage'
  | 'coverage_availability'
  | 'coverage_type';

export type CountryBasic = {
  id: number;
  name: string;
  code: string;
  flag: string;
  map_preview: string;
  description: string;
  data_source: string;
  integration_status: IntegrationStatus;
  date_of_join: string;
  schools_with_data_percentage: number;
  schools_total: number;
  connectivity_availability: ConnectivityAvailability;
  coverage_availability: CoverageAvailability;
  date_schools_mapped: string;
  connectivity_benchmark: number;
  connectivity_benchmark_unit: string;
  latency_benchmark: number;
  latency_benchmark_unit: string;
  uptime_benchmark: number;
  uptime_benchmark_unit: string;
  admin_metadata: AdminMetadataType;
};
export type CoverageBasic = {
  total_coverage_schools: number,
  coverage_schools: {
    "5g_4g": number,
    "3g_2g": number,
    No_coverage: number,
    Unknown: number,
  }
};

export type AdminMetadataType = {
  id: number
  layer_name: string
  name: string
  name_en: string
  description: string
  giga_id_admin: string
  mapbox_id: string
  centroid: number[]
  bbox: number[]
}
export type Country = {
  id: number;
  name: string;
  code: string;
  flag: string;
  map_preview: string;
  description: string;
  data_source: string;
  date_schools_mapped: string;
  statistics: CountryWeeklyStats;
  geometry: GeoJSONGeometry;
  benchmark_metadata: {
    live_layer: Record<string, string>
    default_national_benchmark: Record<string, boolean>
  }
  data_status: {
    week: {
      start_date: string;
      end_date: string;
    },
    month: {
      start_date: string;
      end_date: string;
    }
  }
  admin_metadata: AdminMetadataType;
  admin1_metadata: AdminMetadataType[];
};

export type CountryWeeklyStats = {
  schools_total: number;
  schools_connected: number;
  schools_connectivity_unknown: number;
  schools_connectivity_no: number;
  schools_connectivity_moderate: number;
  schools_connectivity_good: number;
  schools_coverage_unknown: number;
  schools_coverage_no: number;
  schools_coverage_moderate: number;
  schools_coverage_good: number;
  connectivity_speed: number;
  integration_status: IntegrationStatus;
  avg_distance_school: number;
  created: string;
  modified: string;
  connectivity_availability: ConnectivityAvailability;
  coverage_availability: CoverageAvailability;
};

export type CountryGeometry = {
  id: number;
  code: string;
  geometry_simplified: GeoJSONGeometry;
};

export type SchoolSimplified = {
  geopoint: GeoJSONGeometry;
  country_id: number;
  country_integration_status: IntegrationStatus;
};

export type School = {
  id: number;
  name: string;
  geopoint: GeoJSONPoint;
  statistics: {
    num_students: number;
    num_teachers: number;
    num_classroom: number;
    num_latrines: number;
    running_water: boolean;
    electricity_availability: boolean;
    computer_lab: boolean;
    num_computers: number;
    connectivity: boolean | null;
    connectivity_type: string;
    connectivity_speed: number;
    connectivity_latency: number;
    connectivity_availability: number;
    created: string;
    modified: string;
    coverage_type: string;
    coverage_availability: boolean | null;
  };
  gps_confidence: string | null;
  address: string;
  postal_code: string;
  admin1_name: string;
  admin2_name: string;
  timezone: string | null;
  altitude: number;
  email: string | null;
  education_level: string;
  environment: string;
  school_type: string;
  connectivity_status: ConnectionStatus;
  coverage_status: ConnectionStatus;
  is_verified: boolean;
};

export type DailyStats = {
  connectivity_latency: number;
  connectivity_speed: number;
  date: string;
  week: number;
  weekday: WeekDay;
  year: number;
};

export type WeekDay = 1 | 2 | 3 | 4 | 5 | 6 | 7;


export type APIListType<ResultType> = {
  count: number;
  next: string | null;
  previous: string | null;
  results: ResultType[];
}


export interface AdvanceFilterType {
  name: string
  type: string
  description: string
  column_configuration: ColumnConfiguration
  options?: {
    choices?: Choice[]
    placeholder?: string;
    minPlaceholder?: string;
    maxPlaceholder?: string;
    include_none_filter?: boolean
    active_range?: {
      min_value: number
      max_value: number
      min_place_holder: string
      max_place_holder: string
      include_none_filter?: boolean
    }
  }
  query_param_filter: string
}

export interface Choice {
  label: string
  value: string
}

export interface ColumnConfiguration {
  name: string
  label: string
  type: string
  table_name: string
  table_alias: string
  table_label: string
  options?: {
    downcast_aggr_str?: string;
    upcast_aggr_str?: string;
  }
}