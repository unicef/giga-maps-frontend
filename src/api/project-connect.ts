import { createEffect } from 'effector';

import { CoverageStat, LayerType } from '~/@/sidebar/types';
import {
  APIListType,
  AdvanceFilterType,
  ConnectivityStat,
  Country,
  CountryBasic,
  GlobalStats,
  SchoolStatsType,
} from '~/api/types';
import { createRequestFx } from '~/lib/request-fx';
import Controller from '~/lib/request-fx/types';

import { apiBaseUrl, request } from './request-setup';


export const getDatasetUrl = (countryCode: string): string =>
  `${apiBaseUrl}api/locations/countries/${encodeURIComponent(countryCode)}/schools/export-csv-schools/`;

export const getBaseUrl = (url: string): string => `${apiBaseUrl}${url}`;

export const fetchCountryFx = createRequestFx(
  async (countryCode: string, controller?: Controller): Promise<Country> =>
    request({
      url: `api/locations/countries/${encodeURIComponent(countryCode)}/`,
      signal: controller?.getSignal(),
    })
);

export const fetchCountriesFx = createRequestFx(
  async (_, controller?: Controller): Promise<CountryBasic[]> => request({
    url: 'api/locations/countries/',
    signal: controller?.getSignal(),
  })
);

export const fetchLayerListFx = createRequestFx(
  async (_, controller?: Controller): Promise<APIListType<LayerType>> =>
    request({
      url: 'api/accounts/layers/PUBLISHED/?expand=created_by,last_modified_by,published_by&ordering=-last_modified_at',
      signal: controller?.getSignal()
    })
);

export const fetchSchoolPopupDataFx = createRequestFx(
  async ({ query, url }: { query: string; url: string }, controller?: Controller): Promise<SchoolStatsType[]> =>
    request({
      url: `${url}${query}`,
      signal: controller?.getSignal()
    })
);

export const fetchGlobalStatsFx = createRequestFx(
  async ({ query = '' }, controller?: Controller): Promise<GlobalStats> => request({
    url: `api/statistics/global-stat/${query ?? ''}`,
    signal: controller?.getSignal()
  })
);

export const fetchAdvanceFilterFx = createRequestFx(
  async (countryId: number, controller?: Controller): Promise<APIListType<AdvanceFilterType>> => request({
    url: `api/accounts/adv_filters/PUBLISHED/${countryId}/?expand=column_configuration&ordering=name`,
    signal: controller?.getSignal()
  })
);

export const fetchLayerInfoFx = createRequestFx(
  async (url: string, controller?: Controller): Promise<any> =>
    request({
      url,
      signal: controller?.getSignal()
    })
);

export const fetchSchoolLayerInfoFx = createEffect(
  async ({ query, url }: { query: string; url: string }): Promise<SchoolStatsType[]> =>
    fetchLayerInfoFx(`${url}${query}`)
);

export const fetchCountryLiveLayerInfo = createEffect(
  async ({ query, id }: { query: string; id: number | null }): Promise<ConnectivityStat> =>
    fetchLayerInfoFx(`api/accounts/layers/${id}/info/${query}`)
);

export const fetchConnectivityLayerFx = createEffect(
  async ({
    query
  }: {
    query: string;
  },
  ): Promise<ConnectivityStat> => {
    return fetchLayerInfoFx(
      `api/statistics/connectivity/${query}`
    )
  }
);

export const fetchCountryStaticLayerInfo = createEffect(
  async ({ query, id }: { query: string; id: number | null }): Promise<CoverageStat> =>
    fetchLayerInfoFx(`api/accounts/layers/${id}/info/${query}`)
);

export const fetchCoverageLayerFx = createRequestFx(
  async ({ query }: { query: string },
  ): Promise<CoverageStat> => {
    return fetchLayerInfoFx(
      `api/statistics/coverage/${query}`,
    )
  }
);

export const fetchTimePlayerDataFx = createRequestFx(
  async (query: string, controller?: Controller): Promise<any[]> => {
    return request({
      url: `/api/statistics/time-players/${query}`,
      signal: controller?.getSignal(),
    })
  }
);