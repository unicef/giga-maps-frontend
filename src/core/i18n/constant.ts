
export const languages = [
  {
    name: 'English',
    code: 'en',
  },
  // spanish and portuguese
  {
    name: 'Española',
    code: 'es',
  },
  {
    name: 'Português',
    code: 'pt',
  },

]

export const defaultLanguage = 'en'

export const supportedLanguages = languages.map((language) => language.code);