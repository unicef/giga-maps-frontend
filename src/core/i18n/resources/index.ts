import en from './en.json'
import es from './es.json'
import pt from './pt.json'


export { en, es, pt }

