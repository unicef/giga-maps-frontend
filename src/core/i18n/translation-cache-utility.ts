interface CacheConfig {
  useLocalStorage?: boolean;
  prefix?: string;
  defaultExpiration?: number; // in milliseconds
}

interface CacheItem {
  value: string;
  // expiration?: number;
}

type Language = 'en' | 'es' | 'pt' | string; // extendable for more languages

class TranslationCacheUtility {
  private cache: Map<string, CacheItem>;
  private config: CacheConfig;

  constructor(config: CacheConfig = {}) {
    this.cache = new Map();
    this.config = {
      useLocalStorage: config.useLocalStorage || false,
      prefix: config.prefix || 'cache_',
      // defaultExpiration: config.defaultExpiration || 24 * 60 * 60 * 1000, // 24 hours
    };

    if (this.config.useLocalStorage) {
      this.loadFromLocalStorage();
    }
  }

  private getKey(key: string, language: Language): string {
    return `${this.config.prefix}${language}_${key}`;
  }

  set(key: string, value: string, language: Language, expiration?: number): void {
    const cacheKey = this.getKey(key, language);
    const item: CacheItem = {
      value,
      // expiration: expiration || (Date.now() + this.config.defaultExpiration!),
    };

    this.cache.set(cacheKey, item);

    if (this.config.useLocalStorage) {
      this.saveToLocalStorage();
    }
  }

  get(key: string, language: Language): string | undefined {
    const cacheKey = this.getKey(key, language);
    const item = this.cache.get(cacheKey);

    if (item/* && (!item.expiration || item.expiration > Date.now())*/) {
      return item.value;
    }

    if (item) {
      this.cache.delete(cacheKey);
      if (this.config.useLocalStorage) {
        this.saveToLocalStorage();
      }
    }

    return undefined;
  }

  hasKey(key: string, language: Language): boolean {
    const cacheKey = this.getKey(key, language);
    return this.cache.has(cacheKey);
  }

  setMultiple(items: Record<string, string>, language: Language): void {
    Object.entries(items).forEach(([key, value]) => {
      this.set(key, value, language);
    });
  }

  getMultiple(keys: string[], language: Language): Record<string, string | undefined> {
    return keys.reduce((acc, key) => {
      acc[key] = this.get(key, language);
      return acc;
    }, {} as Record<string, string | undefined>);
  }

  remove(key: string, language: Language): void {
    const cacheKey = this.getKey(key, language);
    this.cache.delete(cacheKey);

    if (this.config.useLocalStorage) {
      this.saveToLocalStorage();
    }
  }

  clear(): void {
    this.cache.clear();

    if (this.config.useLocalStorage) {
      this.clearLocalStorage();
    }
  }

  private saveToLocalStorage(): void {
    if (typeof localStorage !== 'undefined') {
      const serialized = JSON.stringify(Array.from(this.cache.entries()));
      localStorage.setItem(this.config.prefix!, serialized);
    }
  }

  private loadFromLocalStorage(): void {
    if (typeof localStorage !== 'undefined') {
      const serialized = localStorage.getItem(this.config.prefix!);
      if (serialized) {
        const items: [string, CacheItem][] = JSON.parse(serialized);
        items.forEach(([key, value]) => this.cache.set(key, value));
      }
    }
  }

  private clearLocalStorage(): void {
    if (typeof localStorage !== 'undefined') {
      localStorage.removeItem(this.config.prefix!);
    }
  }

  // New method to initialize cache with i18n resources
  async initializeWithI18nResources(resources: Record<string, Record<string, string>>): Promise<void> {
    const { en, ...othersResources } = resources;
    const othersLng = Object.keys(othersResources);
    const newResources = {} as Record<string, Record<string, string>>;
    for (let key in en) {
      const newKey = en[key];
      for (let lng of othersLng) {
        if (!newResources[lng]) {
          newResources[lng] = {};
        }
        newResources[lng][newKey] = othersResources[lng][key];
      }
    }
    for (let lang in newResources) { // for each language (resource )
      this.setMultiple(newResources[lang] as Record<string, string>, lang);
    }

    console.log('Cache initialized with i18n resources');
  }
}

export default TranslationCacheUtility;

