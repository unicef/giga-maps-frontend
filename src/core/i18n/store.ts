import i18next from 'i18next';
import { createStore, createEvent } from 'effector';
import { createI18nextIntegration } from '@withease/i18next';
import { languages } from './constant';
import TranslationCacheUtility from './translation-cache-utility';
import * as resources from './resources';

// Event that should be called after application initialization
export const appStarted = createEvent();

// Create Store for i18next instance
const $i18nextInstance = createStore(i18next);

export const languageStore = createI18nextIntegration({
  // Pass Store with i18next instance to the integration
  instance: $i18nextInstance,
  setup: appStarted,
});

export const $SelectedLngObj = languageStore.$language.map((lng) => languages.find((l) => l.code === lng));
export const onLanguageChange = languageStore.changeLanguageFx;

export const translationCache = new TranslationCacheUtility()
translationCache.initializeWithI18nResources(resources)
