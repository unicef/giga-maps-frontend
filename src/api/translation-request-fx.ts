import { createEffect } from 'effector'

import { ConfigNoFn } from '~/lib/request/types';

import { apiBaseUrl, request } from './request-setup'
import { defaultLanguage } from '~/core/i18n/constant';
import { translationCache } from '~/core/i18n/store';

export type ApiRequestType = ConfigNoFn

export const translationRequestFx = createEffect(async (options: ApiRequestType & { lng: string; mapping: [string, string][]; filterMapping: [string, string][] }) => {
  const { headers, baseUrl = `${apiBaseUrl}api/`, mapping, lng, filterMapping, ...rest } = options;
  // if lng is default language, return english translation
  if (lng === defaultLanguage) {
    return {
      data: Object.fromEntries(mapping)
    }
  }
  // if lng is not default language, check if translation is cached
  if (filterMapping?.length === 0) {
    const reverseData = {} as Record<string, string>;
    mapping.forEach((item) => {
      reverseData[item[0]] = translationCache.get(item[1], lng) || item[1];
    })
    return {
      data: reverseData
    }
  }
  return request({
    baseUrl,
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    fn: ({ data }: { data: { translations: [{ text: string }] }[] }) => {
      try {
        const mappedData = {} as any;
        data.forEach((item, index) => {
          mappedData[filterMapping[index][1]] = item.translations[0].text;
        })
        // save to translation cache
        // cache data;
        translationCache.setMultiple(mappedData, lng);

        const reverseData = {} as Record<string, string>;
        mapping.forEach((item) => {
          reverseData[item[0]] = translationCache.get(item[1], lng) || item[1];
        })
        return { data: reverseData };
      } catch (e) {
        console.error(e);
        return { data: {} };
      }
    },
    ...rest
  })
});
