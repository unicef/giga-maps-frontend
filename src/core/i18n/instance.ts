import i18next from "i18next";
import I18nextBrowserLanguageDetector from "i18next-browser-languagedetector";
import { initReactI18next } from "react-i18next";
import { appStarted } from "./store";
import { defaultLanguage, supportedLanguages } from "./constant";
import resources from "./resources";

i18next
  .use(initReactI18next) // passes i18n down to react-i18next
  .use(I18nextBrowserLanguageDetector)
  .init({
    resources: Object.keys(resources).reduce((curr, key) => {
      curr[key] = { translation: resources[key] };
      return curr;
    }, {} as any),
    supportedLngs: supportedLanguages,
    detection: {
      order: ['localStorage', 'navigator'],
      excludeCacheFor: ['localStorage'],
    },
    fallbackLng: defaultLanguage,
    debug: false
  })

// start app;
appStarted();