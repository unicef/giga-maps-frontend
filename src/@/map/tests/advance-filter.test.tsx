import { findByLabelText, fireEvent, render, screen } from "@testing-library/react"

import TextField from "../ui/advanced-filter/text-input"
import FilterPopupContent from "../ui/advanced-filter/filter-popup-content"
import { fetchAdvanceFilterFx, fetchCountryFx } from "~/api/project-connect"
import filterData from "~/tests/data/filter-data"
import RangeTextInput from "../ui/advanced-filter/range-text-input"
import FilterButton from "../ui/advanced-filter/filter"
import { testWrapper } from "~/tests/jest-wrapper"
import "~/core/i18n/instance"
import userEvent from "@testing-library/user-event"

describe('AdvancedFilter', () => {

  beforeEach(() => {
    fetchMock.mockResponse((req) => {
      if (req.url.includes('accounts/adv_filters')) {
        return Promise.resolve(JSON.stringify(filterData))
      } else if (req.url.includes('api/locations/countries/br')) {
        return Promise.resolve(JSON.stringify({
          id: 1,
          name: 'Brazil',
          code: 'br',
        }))
      } else {
        return Promise.resolve(JSON.stringify(null))
      }
    });
  })
  test('should render Advance filter', async () => {

    await fetchAdvanceFilterFx()
    const { asFragment } = render(<FilterPopupContent setOpen={() => { }} />)
    expect(asFragment()).toMatchSnapshot();
  })

  test('should render Advance filter button', async () => {
    await fetchAdvanceFilterFx()
    const { asFragment } = render(testWrapper(<FilterButton />))
    expect(asFragment()).toMatchSnapshot();
  })

  test('should render Advance filter button ', async () => {
    await fetchCountryFx('br')
    await fetchAdvanceFilterFx()
    const { asFragment } = render(testWrapper(<FilterButton />))
    expect(asFragment()).toMatchSnapshot();
  })

  test('should filter button trigger', async () => {
    await fetchCountryFx('br')
    await fetchAdvanceFilterFx()
    render(testWrapper(<FilterButton />))
    const button = (await screen.findByText("Filters")).parentNode as HTMLElement;
    await userEvent.click(button);

    const textInput = (await screen.findByPlaceholderText("Enter building id"))
    await fireEvent.change(textInput, "123");
    // screen.debug();
  })

  test('should render Advance filter', async () => {
    await fetchAdvanceFilterFx()
    const { asFragment } = render(<FilterPopupContent setOpen={() => { }} />)
    expect(asFragment()).toMatchSnapshot();
  })


})

describe.skip('TextField with StyledTextInput', () => {
  const mockOnChange = jest.fn();

  test('should call onChange handler when user types in the input', async () => {
    const { getByPlaceholderText } = render(<TextField
      name="Data Source"
      value=""
      itemKey="data_source"
      type="text"
      description=""
      column_configuration={{ name: '', label: '', type: '', table_name: '', table_alias: '', table_label: '', options: { downcast_aggr_str: '', upcast_aggr_str: '' } }}
      query_param_filter=""
      onChange={mockOnChange}
    />);

    const input = getByPlaceholderText('Enter Data Source')

    await fireEvent.change(input, { target: { value: 'New Value' } });

    expect(mockOnChange).toHaveBeenCalledTimes(1);
  });
});

describe('RangeTextInput Component', () => {
  const mockProps = {
    column_configuration: {
      name: '',
      label: '',
      type: '',
      table_name: '',
      table_alias: '',
      table_label: '',
      options: {
        downcast_aggr_str: '',
        upcast_aggr_str: '',
      }
    },
    options: {
      include_none_filter: true,
    },
    query_param_filter: "",
    name: "No of Teachers",
    type: "range",
    active_countries_range: {
      "30": {
        min_place_holder: "Min (0)",
        max_place_holder: "Max (10K)"
      }
    },
    value: {
      none_range: true,
      value: "15,20",
    },
    itemKey: 'num_computers__range',
    onChange: jest.fn(),
    description: 'test range input',
  };

  it('RangeTextInput input', () => {
    const { container } = render(<RangeTextInput {...mockProps} />);
    expect(container).toMatchSnapshot();
  });

  it('should handle onChange event for min input (valid value)', () => {
    const { getByPlaceholderText } = render(<RangeTextInput {...mockProps} />);
    const minInput = getByPlaceholderText('Min(0)');

    fireEvent.change(minInput, { target: { value: '25' } });

    expect(mockProps.onChange).toHaveBeenCalled();
  });

  it('should handle onChange event for min input (valid value)', () => {
    const { getByPlaceholderText } = render(<RangeTextInput {...mockProps} value={{ "none_range": false, "value": "" }} />);
    const minInput = getByPlaceholderText('Min(0)');

    fireEvent.change(minInput, { target: { value: '0' } });

    expect(mockProps.onChange).toHaveBeenCalledWith('num_computers__range', { "none_range": false, "value": "" });
  });


  it('should handle onChange event for min input (invalid value)', () => {
    mockProps.onChange.mockReset();
    const { getByPlaceholderText } = render(<RangeTextInput {...mockProps} />);
    const minInput = getByPlaceholderText('Min(0)');
    const maxInput = getByPlaceholderText('Max');
    fireEvent.change(minInput, { target: { value: 'abc' } }); // Invalid input
    fireEvent.change(maxInput, { target: { value: 'test' } }); // Invalid input
    expect(mockProps.onChange).toHaveBeenCalled(); // Expect no call
  });

  it('should handle onChange event for min input (empty value)', () => {
    const { getByPlaceholderText } = render(<RangeTextInput {...mockProps} value={{ "none_range": false, "value": "" }} />);
    const maxInput = getByPlaceholderText('Max');

    fireEvent.change(maxInput, { target: { value: '0' } });

    expect(mockProps.onChange).toHaveBeenCalledWith('num_computers__range', { "none_range": false, "value": "" });
  });


  it('should handle onChange event for max input (valid value)', () => {
    const { getByPlaceholderText } = render(<RangeTextInput {...mockProps} />);
    const maxInput = getByPlaceholderText('Max');

    fireEvent.change(maxInput, { target: { value: '25' } });

    expect(mockProps.onChange).toHaveBeenCalledWith('num_computers__range', { "none_range": true, "value": "15,25" });
  });

  it('should handle onChange event on Show null values', () => {
    mockProps.onChange.mockReset();
    const { getByLabelText } = render(<RangeTextInput {...mockProps} />);
    const checkbox = getByLabelText('Show null values');

    fireEvent.click(checkbox);
    expect(mockProps.onChange).toHaveBeenCalled();
  });

  it('should handle onChange event on Show null value with empty min max', () => {
    mockProps.onChange.mockReset();
    const { getByLabelText } = render(<RangeTextInput {...mockProps} value={{ "none_range": true, "value": "" }} />);
    const checkbox = getByLabelText('Show null values');

    fireEvent.click(checkbox);
    expect(mockProps.onChange).toHaveBeenCalled();
  });

});
