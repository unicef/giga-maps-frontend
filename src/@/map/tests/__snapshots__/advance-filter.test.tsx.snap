// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdvancedFilter should render Advance filter 1`] = `
<DocumentFragment>
  .c0 {
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: 1rem;
}

.c0 >h3 {
  font-family: Open Sans;
  font-size: 1rem;
  font-style: normal;
  font-weight: 500;
  line-height: 1.5rem;
}

.c0 button svg {
  fill: !important;
}

.c2 {
  width: 100%;
}

.c2 button {
  width: 50%;
  padding-right: 0;
}

.c1 {
  height: calc(100vh - 9.9rem);
  padding-bottom: 2rem;
}

@media (min-width:769px) {
  .c0 {
    min-width: 18rem;
  }
}

@media (max-width:768px) {
  .c2 {
    position: fixed;
    bottom: 0;
    left: 0;
  }
}

@media (max-width:768px) {
  .c1 {
    max-height: calc(100vh - 13rem);
  }
}

<span
    class="cds--popover"
  >
    <span
      class="cds--popover-content filter-popover-content"
    >
      <div
        class="c0"
      >
        <h3>
          filter-schools-by
        </h3>
        <span
          class="cds--popover-container cds--popover--caret cds--popover--high-contrast cds--popover--bottom-right cds--tooltip cds--icon-tooltip"
        >
          <div
            class="cds--tooltip-trigger__wrapper"
          >
            <button
              aria-labelledby="tooltip-:r0:"
              class="cds--btn--icon-only cds--btn cds--btn--md cds--layout--size-md cds--btn--ghost"
              data-testid="filter-close"
              type="button"
            >
              <svg
                aria-hidden="true"
                fill="currentColor"
                focusable="false"
                height="18"
                preserveAspectRatio="xMidYMid meet"
                viewBox="0 0 32 32"
                width="18"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M17.4141 16L24 9.4141 22.5859 8 16 14.5859 9.4143 8 8 9.4141 14.5859 16 8 22.5859 9.4143 24 16 17.4141 22.5859 24 24 22.5859 17.4141 16z"
                />
              </svg>
            </button>
          </div>
          <span
            aria-hidden="true"
            class="cds--popover"
            id="tooltip-:r0:"
            role="tooltip"
          >
            <span
              class="cds--popover-content cds--tooltip-content"
            >
              Close
            </span>
            <span
              class="cds--popover-caret"
            />
          </span>
        </span>
      </div>
      <form
        aria-label="filter-form"
        class="cds--form"
      >
        <div
          class="scrollbar-container c1 ps"
        >
          <div
            class="ps__rail-x"
            style="left: 0px; top: 0px;"
          >
            <div
              class="ps__thumb-x"
              style="left: 0px; width: 0px;"
              tabindex="0"
            />
          </div>
          <div
            class="ps__rail-y"
            style="top: 0px; left: 0px;"
          >
            <div
              class="ps__thumb-y"
              style="top: 0px; height: 0px;"
              tabindex="0"
            />
          </div>
        </div>
        <div
          class="c2"
        >
          <button
            class="cds--btn cds--btn--secondary"
            type="reset"
          >
            reset
          </button>
          <button
            class="cds--btn cds--btn--primary"
            type="submit"
          >
            apply
          </button>
        </div>
      </form>
    </span>
    <span
      class="cds--popover-caret"
    />
  </span>
</DocumentFragment>
`;

exports[`RangeTextInput Component RangeTextInput input 1`] = `
.c0 {
  padding: 0.5rem 1rem;
}

.c0 .group-input {
  display: flex;
  justify-content: space-between;
  gap: 1rem;
  padding: 0;
}

.c0 .cds--label {
  margin-bottom: 0.5rem;
}

.c2 {
  padding: 0.5rem 1rem;
}

.c3 {
  margin: 0.5rem 0;
}

.c3 .cds--checkbox-label-text {
  font-size: 0.75rem;
}

.c1 button {
  border: none;
  background: inherit;
  cursor: pointer;
}

.c1 .cds--popover-content {
  background: #181818!important;
}

.c1 svg {
  width: 0.75rem;
  height: 0.75rem;
}

.c1 .cds--popover-content {
  max-inline-size: 18rem;
}

<div>
  <div
    class="c0"
  >
    <label
      class="cds--label cds--label--no-margin"
    >
      No of Teachers
      <span
        class="cds--popover-container cds--popover--caret cds--popover--high-contrast cds--popover--bottom cds--tooltip c1"
      >
        <div
          class="cds--tooltip-trigger__wrapper"
        >
          <button
            aria-labelledby="tooltip-:r4:"
            type="button"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              focusable="false"
              height="16"
              preserveAspectRatio="xMidYMid meet"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8.5 11L8.5 6.5 6.5 6.5 6.5 7.5 7.5 7.5 7.5 11 6 11 6 12 10 12 10 11zM8 3.5c-.4 0-.8.3-.8.8S7.6 5 8 5c.4 0 .8-.3.8-.8S8.4 3.5 8 3.5z"
              />
              <path
                d="M8,15c-3.9,0-7-3.1-7-7s3.1-7,7-7s7,3.1,7,7S11.9,15,8,15z M8,2C4.7,2,2,4.7,2,8s2.7,6,6,6s6-2.7,6-6S11.3,2,8,2z"
              />
            </svg>
          </button>
        </div>
        <span
          aria-hidden="true"
          class="cds--popover"
          id="tooltip-:r4:"
          role="tooltip"
        >
          <span
            class="cds--popover-content cds--tooltip-content"
          >
            test range input
          </span>
          <span
            class="cds--popover-caret"
          />
        </span>
      </span>
    </label>
    <div
      class="c2 group-input"
    >
      <div
        class="cds--form-item cds--text-input-wrapper"
      >
        <div
          class="cds--text-input__label-wrapper"
        />
        <div
          class="cds--text-input__field-outer-wrapper"
        >
          <div
            class="cds--text-input__field-wrapper"
          >
            <input
              class="cds--text-input cds--text-input--sm cds--layout--size-sm"
              id="-min-input"
              placeholder="Min(0)"
              title="Min(0)"
              type="number"
              value="15"
            />
            <span
              class="cds--text-input__counter-alert"
              role="alert"
            />
          </div>
        </div>
      </div>
      <div
        class="cds--form-item cds--text-input-wrapper"
      >
        <div
          class="cds--text-input__label-wrapper"
        />
        <div
          class="cds--text-input__field-outer-wrapper"
        >
          <div
            class="cds--text-input__field-wrapper"
          >
            <input
              class="cds--text-input cds--text-input--sm cds--layout--size-sm"
              id="-max-input"
              placeholder="Max"
              title="Max"
              type="number"
              value="20"
            />
            <span
              class="cds--text-input__counter-alert"
              role="alert"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="cds--form-item cds--checkbox-wrapper c3"
    >
      <input
        checked=""
        class="cds--checkbox"
        id="checkbox-num_computers__range"
        style="padding: 0rem 1rem;"
        type="checkbox"
      />
      <label
        class="cds--checkbox-label"
        for="checkbox-num_computers__range"
        title=""
      >
        <span
          class="cds--checkbox-label-text"
        >
          Show null values
        </span>
      </label>
      <div
        class="cds--checkbox__validation-msg"
      />
    </div>
  </div>
</div>
`;
